File: 02/gemini-3.0-pro/typescript/solution.ts
import * as fs from 'fs';

const input = fs.readFileSync('input.txt', 'utf-8').trim();
const ranges = input.split(',');

let totalSum = 0n;

for (const range of ranges) {
    const parts = range.trim().split('-');
    if (parts.length !== 2) continue;

    const start = BigInt(parts[0]);
    const end = BigInt(parts[1]);

    for (let i = start; i <= end; i++) {
        if (isInvalidId(i)) {
            totalSum += i;
        }
    }
}

console.log(`Day 02 Part 1: ${totalSum}`);

let totalSumPart2 = 0n;

for (const range of ranges) {
    const parts = range.trim().split('-');
    if (parts.length !== 2) continue;

    const start = BigInt(parts[0]);
    const end = BigInt(parts[1]);

    for (let i = start; i <= end; i++) {
        if (isInvalidIdPart2(i)) {
            totalSumPart2 += i;
        }
    }
}

console.log(`Day 02 Part 2: ${totalSumPart2}`);

function isInvalidId(id: bigint): boolean {
    const s = id.toString();
    if (s.length % 2 !== 0) return false;

    const halfLen = s.length / 2;
    const firstHalf = s.substring(0, halfLen);
    const secondHalf = s.substring(halfLen);

    return firstHalf === secondHalf;
}

function isInvalidIdPart2(id: bigint): boolean {
    const s = id.toString();
    const len = s.length;

    for (let patternLen = 1; patternLen <= Math.floor(len / 2); patternLen++) {
        if (len % patternLen !== 0) continue;

        const pattern = s.substring(0, patternLen);
        const repetitions = len / patternLen;

        if (pattern.repeat(repetitions) === s) {
            return true;
        }
    }

    return false;
}

File: 02/gemini-3.0-pro/python/solution.py
with open("input.txt") as f:
    input_data = f.read().strip()

ranges = input_data.split(',')
total_sum = 0

for r in ranges:
    parts = r.strip().split('-')
    if len(parts) != 2:
        continue

    start = int(parts[0])
    end = int(parts[1])

    for i in range(start, end + 1):
        s = str(i)
        if len(s) % 2 != 0:
            continue

        half_len = len(s) // 2
        first_half = s[:half_len]
        second_half = s[half_len:]

        if first_half == second_half:
            total_sum += i

print(f"Day 02 Part 1: {total_sum}")

total_sum_part2 = 0

for r in ranges:
    parts = r.strip().split('-')
    if len(parts) != 2:
        continue

    start = int(parts[0])
    end = int(parts[1])

    for i in range(start, end + 1):
        s = str(i)
        length = len(s)
        is_invalid = False

        for pattern_len in range(1, length // 2 + 1):
            if length % pattern_len != 0:
                continue

            pattern = s[:pattern_len]
            repetitions = length // pattern_len

            if pattern * repetitions == s:
                is_invalid = True
                break

        if is_invalid:
            total_sum_part2 += i

print(f"Day 02 Part 2: {total_sum_part2}")
